# Release 3.3.0

Hey Scout-users! It's time for a new update with quite a few features this time :D

Please read below for more information.

## New features!

1. Cases are now listed by status in separate groups
2. Upload of gene panel export
3. Dynamic gene panel is back! You can now upload a list of HGNC symbols to do a temporary search
4. Scout will now display exon as well as intron number in the Sanger email
5. We've added a new "Users" view where you can see an overview of all Scout users - check it out!
6. You can now filter variants on local observations in the SNV view and they will show up in the popup
7. ACMG evaluation for variants is now available
8. Scout will now display if a variant is in the PAR region or not
9. You can now update the default gene panels directly in Scout - no need to rerun!
10. We've added labels to the variant tags according to proposed example from CMMS
11. If variant is in an autozygocity block it will now be displayed

## Bugfixes

1. A warning is now displayed is you search for a gene symbol that doesn't exist in Scout
2. When filtering on frequencies, Scout now handles input as "0,04" and "0.04"
3. We will now use coordinates instead of gene identifiedr when looking for overlapping SV/SNV variants
4. Scout now correctly displays a notice if a variant has been commented on
5. CLINSIG annotations are now properly displayed!
6. We've fixed display of expected inheritance model on variant page

## Future updates

We have prepared additional fearures which will not immidiately show up in Scout but might require e.g. an update to MIP.

1. New cancer view: we've added rudementary support for display cancer variants in Scout as a separate view much like the SV view.
2. We've added support for displaying information from Peddy such as confirmed parenthood, sex, and predicted ancestry

Happy summer!
