{% macro gtcall_panel(variant) %}
  <div class="card panel-default">
    <div class="panel-heading">
      GT call
      {% if variant.is_par %}
        <span class="badge bg-info">PAR</span>
      {% endif %}
    </div>
    <div class="card-body">
    <div class="table-responsive">
      <table class="table table-bordered">
        <thead class="thead table-light">
          <tr>
            <th rowspan="2">Sample</th>
            <th rowspan="2">Genotype (GT)</th>
            <th colspan="2" title="Unfiltered count of reads that support a given allele.">Allele depth (AD)</th>
            {% if variant.category == "cancer" %}
              <th rowspan="2" colspan="1" title="Variant Allele Frequency.">Variant Allele Frequency (VAF)</th>
            {% elif variant.category == "sv" %}
              <th rowspan="2" colspan="1" title="SV caller specific quality score. Note different scales for different callers.">SV quality</th>
            {% elif variant.category == "str" %}
              <th rowspan="2" colspan="1">ExpansionHunter support</th>
            {% else %}
	      {% if variant.chromosome in ["MT","M"] %}
                <th rowspan="2" colspan="1" title="Variant Allele Frequency.">Variant Allele Frequency (VAF)</th>
	      {% endif %}
              <th rowspan="2" colspan="1" title="Phred-scaled confidence that the true genotype is the one provided in GT (max=99).">Genotype quality (GQ)</th>
            {% endif %}
            <tr>
              <th>Reference</th>
              <th>Alternative</th>
            </tr>
          </tr>
        </thead>
        <tbody>
          {% for sample in variant.samples %}
            <tr {{ 'class="danger"' if sample.is_affected }}>
              <td>{{ sample.display_name }}</td>
              <td class="text-center">{{ sample.genotype_call }}</td>
              {% if sample.allele_depths %}
                  {% for number in sample.allele_depths %}
                    <td class="text-end">
                      {% if number == -1 %}
                        <small>N/A</small>
                      {% else %}
                        {{ number }}
                      {% endif %}
                    </td>
                  {% endfor %}
              {% else %}
                  <td class="text-end"><small>N/A</small></td>
                  <td class="text-end"><small>N/A</small></td>
              {% endif %}

              {% if variant.category == "snv" and variant.chromosome in ["MT","M"] %}
                <td>
                  {% if sample.alt_frequency and sample.alt_frequency != -1 %}
                    {{ (100*sample.alt_frequency)|round(2) }}%
                  {% else %}
		    N/A
                  {% endif %}
                </td>
              {% endif %}
              {% if variant.category == "cancer" %}
                {% set cancer_var = namespace() %}
                {% if variant.tumor and sample.sample_id == variant.tumor.ind_id %}
                  {% set cancer_var.vaf = variant.tumor.alt_freq|round(4) %}
                {% elif variant.normal and sample.sample_id == variant.normal.ind_id %} <!--normal-->
                  {% set cancer_var.vaf = variant.normal.alt_freq|round(4) %}
                {% else %}
                  {% set cancer_var.vaf = "N/A" %}
                {% endif %}
                <td class="text-end">{{ cancer_var.vaf }}</td>
              {% else %}
                <td class="text-end">
                {% if variant.category == "sv" %}
                  {% if sample.genotype_quality not in ["None", None, "-1", -1] %}
                    {{ sample.genotype_quality }}
                  {% else %}
                    <small>N/A</small>
                  {% endif %} (<small>VQ</small>
                  {% if variant.quality not in ["None", None, "-1", -1] %}
                    {{variant.quality}}
                  {% else %}
                    <small>N/A</small>
                  {% endif%})
                {% elif variant.category == "str" %}
                    <small>{{ sample.so }}</small>
                {% else %}
                  {{ sample.genotype_quality }}
                {% endif %}
                </td>
              {% endif %}
            </tr>
          {% endfor %}
        </tbody>
      </table>
    </div>
    <div class="panel-footer">
      {% for name, caller in variant.callers %}
        <span class="badge bg-secondary">{{ name }}: {{ caller }}</span>
      {% endfor %}
    </div>
    </div>
  </div>
{% endmacro %}

{% macro frequencies(variant) %}
  <div class="card panel-default">
    <div class="panel-heading">Frequencies</div>
    <div class="card-body">
      <table class="table">
        <thead class="thead table-light">
          <tr>
            <th scope="col">Source</th>
            <th scope="col">Frequency</th>
          </tr>
        </thead>
        <tbody>
          {% for freq_name, value, link in variant.frequencies %}
          <tr>
            <td>
              {% if link %}
                <a href="{{ link }}" target="_blank" rel="noopener" referrerpolicy="no-referrer">{{ freq_name }}</a>
              {% else %}
                {{ freq_name }}
              {% endif %}
            </td>
            <td>
              {% if value %}
                <span class="badge bg-secondary">{{ value|human_decimal }}</span>
              {% else %}
                -
              {% endif %}
            </td>
          </tr>
          {% endfor %}
        </tbody>
      </table>
    </div>
  </div>
{% endmacro %}

{% macro observations_panel(variant, observations, case) %}
  <div class="card panel-default">
    <div class="panel-heading d-flex justify-content-between">
      <a href="https://github.com/moonso/loqusdb" target="_blank">Local observations</a>
      <span data-bs-toggle="tooltip" title="Nr of observations is the total number of occurrences in the loqusdb instance. Missing links on greyed out case names for shared local case variants might be caused by variants not loaded after a rerun or inexact SV matching. Only names of collaborator cases are shown.">?</span>
    </div>
    <div class="card-body">
      <table class="table">
        <thead class="thead table-light">
          <tr>
            <th scope="col">ID</th>
            <th scope="col">Nr obs.</th>
            <th scope="col">Nr homo.</th>
            <th scope="col">Total nr.</th>
          </tr>
        </thead>
        <tbody>
          {% for loqusid, obs in observations.items() %}
              <tr>
                <td>{{ loqusid }}</td>
               <td>{{ obs.observations|default('N/A') }}</td>
               <td>{{ obs.homozygote|default('N/A') }}</td>
               <td>{{ obs.total|default('N/A') }}</td>
             </tr>
           {% if obs.cases %}
             <tr>
               <td colspan="4">
                     {% for data in obs.cases %}
                       {% if data.variant and data.variant.category == "snv"%}
                         <a class="badge rounded-pill bg-light text-dark" target="_blank" href="{{ url_for('variant.variant', institute_id=data.case.owner, case_name=data.case.display_name, variant_id=data.variant._id) }}">{{ data.case.display_name }}</a>
                       {% elif data.variant and data.variant.category == "sv"%}
                         <a class="badge rounded-pill bg-light text-dark" target="_blank" href="{{ url_for('variant.sv_variant', institute_id=data.case.owner, case_name=data.case.display_name, variant_id=data.variant._id) }}">{{ data.case.display_name }}</a>
                       {% else %}
                         <span class="ml-3 badge rounded-pill bg-light text-dark">{{ data.case.display_name }}</span>
                       {% endif %}
                     {% endfor %}
                     {% if obs.cases|length > 10 %}
                       + more
                     {% endif %}
                     {% if obs.observations == 1 %}
                       Observed only in this case
                     {% endif %}
               </td>
             </tr>
           {% endif %}
          {% endfor %}
        </tbody>
      </table>
    </div>
  </div>
{% endmacro %}

{% macro old_observations(variant, obs_date) %}
{% if "SHOW_OBSERVED_VARIANT_ARCHIVE" in config and config["SHOW_OBSERVED_VARIANT_ARCHIVE"] == true %}
  <div class="card panel-default">
    <div class="panel-heading">
      <a href="https://github.com/moonso/loqusdb" target="_blank" data-bs-toggle="tooltip" title="Local observations annotated from an archive copy of the loqusdb local frequency database. Only variants above a quality threshold are included, and each case is included only once. {% if variant.local_obs_old_desc %} {{variant.local_obs_old_desc}} {% endif %}">Local observations (archive {{ obs_date or variant.local_obs_old_date or "version"}})</a>
    </div>
    <div class="card-body">
      <table class="table">
        <thead class="thead table-light">
          <tr>
            <th scope="col">Archive</th>
            <th scope="col">Nr obs.</th>
            <th scope="col">Nr homo.</th>
            <th scope="col">Frequency</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>RD</td>
            <td>{{ variant.local_obs_old|default('N/A') }}</td>
            <td>{{ variant.local_obs_hom_old|default('N/A') }}</td>
            <td>{% if variant.local_obs_old_freq %} {{variant.local_obs_old_freq|round(6)}} {% elif variant.local_obs_old_nr_cases and variant.local_obs_old %} {{ (variant.local_obs_old/variant.local_obs_old_nr_cases)|round(5) }} {% elif variant.local_obs_old_nr_cases %} 0 / {{variant.local_obs_old_nr_cases}} {% else %} N/A {% endif %}</td>
          </tr>
          {% if variant.category in ['cancer', 'cancer_sv'] %}
            <tr>
              <td>Cancer Germline</td>
              <td>{{ variant.local_obs_cancer_germline_old|default('N/A') }}</td>
              <td>{{ variant.local_obs_cancer_germline_hom_old|default('N/A') }}</td>
              <td>{% if variant.local_obs_cancer_germline_old_freq %} {{variant.local_obs_cancer_germline_old_freq|round(6)}} {% elif variant.local_obs_cancer_germline_old_nr_cases and variant.local_obs_cancer_germline_old %} {{ (variant.local_obs_cancer_germline_old/variant.local_obs_cancer_germline_old_nr_cases)|round(5) }} {% elif variant.local_obs_cancer_germline_old_nr_cases %} 0 / {{variant.local_obs_cancer_germline_old_nr_cases}} {% else %} N/A {% endif %}</td>
            </tr>
            <tr>
              <td>Cancer Somatic</td>
              <td>{{ variant.local_obs_cancer_somatic_old|default('N/A') }}</td>
              <td>{{ variant.local_obs_cancer_somatic_hom_old|default('N/A') }}</td>
              <td>{% if variant.local_obs_cancer_somatic_old_freq %} {{variant.local_obs_cancer_somatic_old_freq|round(6)}} {% elif variant.local_obs_cancer_somatic_old_nr_cases and variant.local_obs_cancer_somatic_old %} {{ (variant.local_obs_cancer_somatic_old/variant.local_obs_cancer_somatic_old_nr_cases)|round(5) }} {% elif variant.local_obs_cancer_somatic_old_nr_cases %} 0 / {{variant.local_obs_cancer_somatic_old_nr_cases}} {% else %} N/A {% endif %}</td>
            </tr>
          {% endif %}
        </tbody>
      </table>
    </div>
  </div>
{% endif %}
{% endmacro %}

{% macro severity_list(variant) %}
  <div class="card panel-default">
    <div class="panel-heading">Severity</div>
    <div class="card-body">
      <ul class="list-group">
        <li class="list-group-item">
          <a href="https://sift.bii.a-star.edu.sg/www/SIFT_help.html" target="_blank">SIFT</a>
          <span class="float-end">
            {% if variant.sift_predictions  %}
              {{ variant.sift_predictions|join(', ') }}</span>
            {% else %}
               {{ "-" }}
            {% endif %}
          </span>
        </li>
        <li class="list-group-item">
          <a href="https://sites.google.com/site/revelgenomics/about" target="_blank" data-bs-toggle="tooltip"
            title="An ensemble score based on 13 individual scores for predicting the pathogenicity of missense variants. Scores range from 0 to 1. The larger the score the more likely the SNP has damaging effect">REVEL score</a>
          <span class="float-end">
          {% if variant.revel  %}
             {{ variant.revel }}
          {% else %}
             {{ "-" }}
          {% endif %}
          </span>
        </li>
        <li class="list-group-item">
          <a href="http://database.liulab.science/dbNSFP" target="_blank" data-bs-toggle="tooltip"
             title="REVEL scores were ranked among all REVEL scores in dbNSFP. The REVEL rank score is the ratio of the rank of the score over the total number of REVEL scores in dbNSFP">REVEL rank score</a>
          <span class="float-end">
          {% if variant.revel_score  %}
             {{ variant.revel_score }}
          {% else %}
             {{ "-" }}
          {% endif %}
          </span>
        </li>
        <li class="list-group-item">
          <a href="http://genetics.bwh.harvard.edu/pph2/" target="_blank">Polyphen</a>
          <span class="float-end">
          {% if variant.polyphen_predictions %}
            {{ variant.polyphen_predictions|join(', ') }}
          {% else %}
             {{ "-" }}
          {% endif %}
          </span>
        </li>
        <li class="list-group-item">
          SPIDEX
          <span class="float-end">{{ variant.spidex_human }}</span>
        </li>
        <li class="list-group-item">
          <a href="{{ variant.spliceai_link }}" target="_blank" rel="noopener">SpliceAI</a> <a href="https://github.com/Illumina/SpliceAI" target="_blank" rel="noopener">DS max</a>
          <span class="float-end control-label" data-bs-toggle="tooltip" data-bs-html="true" data-bs-placement="bottom"
              title="<strong>
                {% for entry in variant.spliceai_scores %}
                  {% if entry is not none %}
                    SpliceAI highest delta score {{ entry }} </strong> at position {{ variant.spliceai_positions[loop.index0]}} relative to this variant.
                    {% if variant.spliceai_predictions[loop.index0] %}
                      <br>All scores and positions(relative to variant):<br>
                      {{ variant.spliceai_predictions[loop.index0] }} <br>
                    {%  endif %}
                    <br>
                  {% else %}
                    No SpliceAI positions annotated for this variant.
                  {% endif %}
                {% endfor %}
              ">
            {{ variant.spliceai_scores|join(', ') or "-"}}
          </span>
        </li>
      </ul>
    </div>
  </div>
{% endmacro %}

{% macro conservations(variant) %}
  <div class="card panel-default">
    <div class="panel-heading">Conservation</div>
    <div class="card-body">
      <ul class="list-group">
        <li class="list-group-item">
          PHAST
          <span class="float-end">{{ variant.phast_conservation|join(', ') or '-' }}</span>
        </li>
        <li class="list-group-item">
          GERP
          <span class="float-end">{{ variant.gerp_conservation|join(', ') or '-' }}</span>
        </li>
        <li class="list-group-item">
          phyloP
          <span class="float-end">{{ variant.phylop_conservation|join(', ') or '-' }}</span>
        </li>
      </ul>
    </div>
  </div>
{% endmacro %}


{% macro mappability_list(variant) %}
  {% set superdups_fracmatches = [] %}
  {% for gene in variant.genes %}
    {% for transcript in gene.transcripts %}
      {% if transcript.superdups_fracmatch %}
        {% for superdup in transcript.superdups_fracmatch %}
          {% do superdups_fracmatches.append(superdup) %}
        {% endfor %}
      {% endif %}
    {% endfor %}
  {% endfor %}
  {% set superdups_fracmatches = superdups_fracmatches|sort %}
  {% if superdups_fracmatches %}
    <li class="list-group-item">
      <span>mapping to {{superdups_fracmatches|length}} segm. dups. (min matching:{{ superdups_fracmatches|first|float|round(3) }}, max matching:{{ superdups_fracmatches|last|float|round(3) }})</span>
    </li>
  {% else %}
    <li class="list-group-item">
      <p class="card-text">-</p>
    </li>
  {% endif %}
{% endmacro %}

{% macro mappability(variant) %}
  <div class="card panel-default">
    <div class="panel-heading">Mappability (<a href="http://genome.ucsc.edu/cgi-bin/hgTrackUi?g=genomicSuperDups" target="_blank">fracMatch</a>)</div>
    <div class="panel-body">
      <ul class="list-group">
        {{ mappability_list(variant) }}
      </ul>
    </div>
  </div>
{% endmacro %}
