# Evidence of pathogenecity

## Very Strong

### PVS1
Null variant (nonsense, frameshift, canonical ±1 or 2 splice sites, initiation
 codon, single or multiexon deletion) in a gene where LOF is a known mechanism of disease

**Caveats:**

 - Beware of genes where LOF is not a known disease mechanism (e.g., GFAP, MYH7)
 - Use caution interpreting LOF variants at the extreme 3′ end of a gene
 - Use caution with splice variants that are predicted to lead to exon skipping but leave the remainder of the  protein intact
 - Use caution in the presence of multiple transcripts

## Strong

### PS1
Same amino acid change as a previously established pathogenic variant regardless of nucleotide change
*Example:* Val→Leu caused by either G>C or G>T in the same codon

**Caveat:**
Beware of changes that impact splicing rather than at the amino acid/protein level

### PS2
De novo (both maternity and paternity confirmed) in a patient with the
disease and no family history.

**Note:** Confirmation of paternity only is insufficient. Egg donation,
surrogate motherhood, errors in embryo transfer, and so on, can contribute
to nonmaternity.

### PS3
Well-established in vitro or in vivo functional studies supportive of a
damaging effect on the gene or gene product.

**Note:** Functional studies that have been validated and shown to be
reproducible and robust in a clinical diagnostic laboratory setting are
considered the most well established.

### PS4
The prevalence of the variant in affected individuals is significantly
increased compared with the prevalence in controls.

**Note 1:** Relative risk or OR, as obtained from case–control studies,
is >5.0, and the confidence interval around the estimate of relative risk
or OR does not include 1.0. See the article for detailed guidance.

**Note 2:** In instances of very rare variants where case–control studies
may not reach statistical significance, the prior observation of the variant
in multiple unrelated patients with the same phenotype, and its absence in
controls, may be used as moderate level of evidence.

## Moderate

### PM1
Located in a mutational hot spot and/or critical and well-established
functional domain (e.g., active site of an enzyme) without benign variation.

### PM2
Absent from controls (or at extremely low frequency if recessive) (Table 6)
in Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation
Consortium

**Caveat:** Population data for insertions/deletions may be poorly called
by next-generation sequencing.

### PM3
For recessive disorders, detected in trans with a pathogenic variant

**Note:** This requires testing of parents (or offspring) to determine phase.

### PM4
Protein length changes as a result of in-frame deletions/insertions in a
nonrepeat region or stop-loss variants.

### PM5
Novel missense change at an amino acid residue where a different missense
change determined to be pathogenic has been seen before.

**Example:** Arg156His is pathogenic; now you observe Arg156Cys
**Caveat:** Beware of changes that impact splicing rather than at the amino acid/protein level.
PM6 Assumed de novo, but without confirmation of paternity and maternity

## Supporting

### PP1
Cosegregation with disease in multiple affected family members in a gene
definitively known to cause the disease.

**Note:** May be used as stronger evidence with increasing segregation data

### PP2
Missense variant in a gene that has a low rate of benign missense variation
and in which missense variants are a common mechanism of disease.

### PP3
Multiple lines of computational evidence support a deleterious effect on the
gene or gene product  (conservation, evolutionary, splicing impact, etc.).

**Caveat:** Because many in silico algorithms use the same or very similar
input for their predictions, each algorithm should not be counted as an
independent criterion. PP3 can be used only once in any evaluation of a variant.

## PP4
Patient’s phenotype or family history is highly specific for a disease with
a single genetic etiology.

## PP5
Reputable source recently reports variant as pathogenic, but the evidence is
not available to the laboratory to perform an independent evaluation.

# Evidence of benign impact

## Stand-alone

### BA1
Allele frequency is >5% in Exome Sequencing Project, 1000 Genomes Project,
or Exome Aggregation Consortium.

## Strong

### BS1
Allele frequency is greater than expected for disorder (see Table 6)

### BS2
Observed in a healthy adult individual for a recessive (homozygous),
dominant (heterozygous), or X-linked (hemizygous) disorder, with full
penetrance expected at an early age.

### BS3
Well-established in vitro or in vivo functional studies show no damaging
effect on protein function or splicing.

### BS4
Lack of segregation in affected members of a family

**Caveat:** The presence of phenocopies for common phenotypes (i.e., cancer,
epilepsy) can mimic lack of segregation among affected individuals. Also,
families may have more than one pathogenic variant contributing to an
autosomal dominant disorder, further confounding an apparent lack of segregation.

## Supporting

### BP1
Missense variant in a gene for which primarily truncating variants are known
to cause disease.

### BP2
Observed in trans with a pathogenic variant for a fully penetrant dominant
gene/disorder or observed in cis with a pathogenic variant in any inheritance
pattern.

### BP3
In-frame deletions/insertions in a repetitive region without a known function.

### BP4
Multiple lines of computational evidence suggest no impact on gene or gene
product (conservation, evolutionary, splicing impact, etc.).

**Caveat:** Because many in silico algorithms use the same or very similar
input for their predictions, each algorithm cannot be counted as an
independent criterion. BP4 can be used only once in any evaluation of a variant.

### BP5
Variant found in a case with an alternate molecular basis for disease.

### BP6
Reputable source recently reports variant as benign, but the evidence is not
available to the laboratory to perform an independent evaluation.

### BP7
A synonymous (silent) variant for which splicing prediction algorithms predict
no impact to the splice consensus sequence nor the creation of a new splice
site AND the nucleotide is not highly conserved.

# Modifying criteria strength

The ACMG criteria allow for modifying the weight of different criteria based on the
strength of the evidence, or certain modifying information, with a lab manager's professional judgement.
A common, well documented application is to downgrade `PVS1` variants (Tayoun et al, Human mutation 2019). Another would be the application of a modified segregation in the family based on different levels of evidence.
Or the recognition that variants sometimes do not meet the strict frequency criteria of `PM2`, `BA1` or `BS1` but that a tendency could warrant a modified
`PM2_Supporting` or `BS1_Supporting` (Harrison et al Curr Prot Hum Genet 2019, PMID: 31479589)[https://pubmed.ncbi.nlm.nih.gov/31479589/].

We apply the modification nomenclature introduced by the [ClinGen variant interpretation working group in 2017](https://www.clinicalgenome.org/site/assets/files/3459/svi_criteria_nomenclature_recommendation_v1.pdf)
Compare e.g. the [Baylor approved criteria](https://www.baylorgenetics.com/variant-classification/).

