{% macro variant_gene_symbols_cell(variant) %}
  <div class="d-flex justify-content-between align-items-center">
    <div>
      {% if variant.hgnc_symbols|length >= 5 %}
        <span class="badge bg-secondary mr-3 text-white">{{ variant.hgnc_symbols|length }}</span>
        <a class="mr-3" data-bs-toggle="collapse" href="#_genes_{{variant._id}}" role="button" aria-expanded="false" aria-controls="_{{variant._id}}">[...]</a>
        <div class="collapse" id="_genes_{{variant._id}}">
          <span class="text-body">
          {% for symbol in variant.hgnc_symbols|sort %}
            {{symbol}}<br>
          {% endfor %}
          </span>
        </div>
      {% else %}
        <div>
          {% if variant.hgnc_symbols %}
            <span class="badge bg-secondary mr-3 text-white">{{ variant.hgnc_symbols|length }}</span>
          {% endif %}
          <span class="text-body">{{variant.hgnc_symbols|sort|join(", ")}}</span>
        </div>
      {% endif %}
    </div>
  </div>
{% endmacro %}

{% macro fusion_variant_gene_symbols_cell(variant) %}
<div class="align-items-center">
  {% if variant.genes %}
    {% for gene in variant.genes %}
       <a data-bs-toggle="tooltip" data-bs-html="true" title="
         <div>
              <strong>{{ gene.hgnc_symbol or gene.hgnc_id }}</strong>:  {{ gene.description }}
         </div>"
        href="{{ url_for('genes.gene', hgnc_id=gene.hgnc_id) }}">{{ gene.hgnc_symbol or gene.hgnc_id }}
        </a> <br>
    {% endfor %}
  {% else %}
    {% for symbol in variant.fusion_genes %}
      {{ symbol }} <br>
    {% endfor %}
  {% endif %}
</div>
{% endmacro %}

{% macro variant_funct_anno_cell(variant) %}
  <div class="d-flex justify-content-between align-items-center">
    <div>
      {% if variant.category == "fusion" %}
        {{ variant.frame_status|lower if variant.frame_status else "-" }}
      {% else %}
        {% if variant.functional_annotations|length >= 5 %}
          <a class="mr-3" data-bs-toggle="collapse" href="#_functanno_{{variant._id}}" role="button" aria-expanded="false" aria-controls="_{{variant._id}}">{% if variant.missing_data %}[first 30 genes]{% else %}[...]{% endif %}</a>
          <div class="collapse" id="_functanno_{{variant._id}}">
            <span class="text-body">
            {% for annotation in variant.functional_annotations|sort %}
              {{ annotation|replace("_", " ") }}<br>
            {% endfor %}
            </span>
          </div>
        {% else %}
          <div>
            <span class="text-body">{{ variant.functional_annotations|sort|join(", ")|replace("_", " ") }}</span>
          </div>
        {% endif %}
      {% endif %}
    </div>
  </div>
{% endmacro %}

{% macro variant_region_anno_cell(variant) %}
  <div class="d-flex justify-content-between align-items-center">
    <div>
      {% if variant.functional_annotations|length >= 5 %}
        <a class="mr-3" data-bs-toggle="collapse" href="#_regionanno_{{variant._id}}" role="button" aria-expanded="false" aria-controls="_{{variant._id}}">{% if variant.missing_data %}[first 30 genes]{% else %}[...]{% endif %}</a>
        <div class="collapse" id="_regionanno_{{variant._id}}">
          {% for annotation in variant.region_annotations|sort %}
            {{ annotation }}<br>
          {% endfor %}
        </div>
      {% else %}
        <div>
          {{ variant.region_annotations|sort|join(", ") }}
        </div>
      {% endif %}
    </div>
  </div>
{% endmacro %}

{% macro gene_cell(variant) %}
  <div class="align-items-center">
    {% if variant.category == "cancer" or variant.category == "sv_cancer" %}
      <a data-bs-toggle="tooltip" data-bs-html="true" title="
        <div>
          <div>
            <strong>{{ variant.first_rep_gene.hgnc_symbol }}</strong>: {{ variant.first_rep_gene.description }}
          </div>
          {% if variant.first_rep_gene.inheritance %}
            <div>
              <strong>Models</strong>: {{ variant.first_rep_gene.inheritance|join(',') }}
            </div>
          {% endif %}
        {% if variant.first_rep_gene.phenotypes %}
          <div><strong>OMIM disease</strong>
          {% for disease in variant.first_rep_gene.phenotypes %}
            <div>
                {{ disease.description }}
            </div>
          {% endfor %}
          </div>
        {% endif %}
        </div>"
        {% if variant.first_rep_gene %}
          href="{{ url_for('genes.gene', hgnc_id=variant.first_rep_gene.hgnc_id) }}"
        {% endif %}>
        {{ variant.first_rep_gene.hgnc_symbol or variant.first_rep_gene.hgnc_id }}
        {% if variant.secondary_gene %}
          <span class="text-muted">
            ({{ variant.second_rep_gene.hgnc_symbol or variant.second_rep_gene.hgnc_id }})
          </span>
        {% endif %}
      </a>

      {% set panel_count = variant.case_panels|rejectattr('removed')|list|length %}
      {% if panel_count > 0%}
          <a
            class="badge bg-secondary text-white"
            data-bs-toggle="popover"
            data-bs-html="true"
            data-bs-trigger="hover click"
            title="Overlapping gene panels"
            data-bs-content="{% for panel in variant.case_panels|sort(attribute='panel_name',case_sensitive=False)|rejectattr('removed') %}
              {{ panel.panel_name|safe }}<br>
              {% endfor %}"
            >{{panel_count}}
          </a>
      {% endif %}
    {% else %}
      {% for gene in variant.genes %}
        <a data-bs-toggle="tooltip" data-bs-html="true" title="
        <div>
          <div>
            <strong>{{ gene.hgnc_symbol }}</strong>: {{ gene.description }}
          </div>
          {% if gene.inheritance %}
            <div>
              <strong>Models</strong>: {{ gene.inheritance|join(',') }}
            </div>
          {% endif %}
          {% if gene.phenotypes %}
            <div><strong>OMIM disease</strong>
            {% for disease in gene.phenotypes %}
              <div>
                  {{ disease.description }}
              </div>
            {% endfor %}
            </div>
          {% endif %}
        </div>"
          href="{{ url_for('genes.gene', hgnc_id=gene.hgnc_id) }}">{{ gene.hgnc_symbol or gene.hgnc_id }}
        </a>
      {% endfor %}


      {% set panel_count = variant.case_panels|rejectattr('removed')|list|count %}
      {% if panel_count > 0 %}
          <a
            class="badge bg-secondary text-white"
            data-bs-toggle="popover"
            data-bs-html="true"
            data-bs-trigger="hover click"
            data-bs-content="{% for panel in variant.case_panels|sort(attribute='panel_name',case_sensitive=False)|rejectattr('removed') %}
              {{ panel.panel_name|safe }}<br>
              {% endfor %}"
            title="Overlapping gene panels">{{panel_count}}
          </a>
      {% endif %}
    {% endif %}
  </div>
{% endmacro %}

{% macro observed_cell_general(variant) %}
  <div data-bs-toggle="tooltip" data-bs-placement="right" data-bs-html="true" title="
    <div class='text-start'>
      <div>
        <strong>Local (arch.)</strong>:
        {% if variant.category in ['snv', 'sv'] %}
          {{ variant.local_obs_old or 0 }}
        {% elif variant.category in ['cancer', 'cancer_sv'] %}
          germline:{{ variant.local_obs_cancer_germline_old or 0}}, somatic:{{variant.local_obs_cancer_somatic_old or 0}}, rare-disease: {{ variant.local_obs_old or 0}},
        {% endif %} obs.
      </div>
      {% if variant.mitomap_associated_diseases %}
        <strong>MITOMAP</strong>: {{ variant.mitomap_associated_diseases }}<br>
      {% endif %}
      {% if variant.cosmic_ids %}
      <strong>COSMIC</strong>: {{variant.cosmic_ids|count}} obs
      <br>
      {% endif %}
      {% if variant.clingen_cgh_pathogenic %}
          <strong>ClinGen CGH Pathogenic</strong>: {{ variant.clingen_cgh_pathogenic }}<br>
      {%  endif %}
      {% if variant.clinsig_human %}
        <strong>ClinVar</strong>:
          <ul>
          {% for clinsig in variant.clinsig_human %}
            <li>
              <strong>{{clinsig.human}}</strong>
              {% if clinsig.revstat %}{{ clinsig.revstat }} {% endif %}
              {% if clinsig.accession %}{{ clinsig.accession }}{% else %}no accnr{% endif %}
            </li>
          {% endfor %}
          </ul>
      {% endif %}

    </div>
  ">
  {% if variant.clinsig_human %}
    <span class="badge bg-secondary">ClinVar</span>
    <br>
  {% endif%}
  {% if variant.cosmic_ids %}
    <span class="badge bg-secondary">COSMIC</span>
    <br>
  {% endif %}
  {% if variant.mitomap_associated_diseases %}
    <span class="badge bg-secondary">MITOMAP</span>
    <br>
  {% endif %}
  {% if variant.clingen_cgh_pathogenic %}
    <span class="badge bg-secondary">CGH Pathogenic</span><br>
  {% endif %}
  {% if variant.local_obs_old or variant.local_obs_cancer_germline_old or variant.local_obs_cancer_somatic_old %}
    <span class="badge bg-secondary">Local</span>
    <br>
  {% endif %}
  </div>
{% endmacro %}

{% macro frequency_cell_general(variant) %}
  <div class="d-flex justify-content-between" data-bs-toggle="tooltip" data-bs-placement="left" data-bs-html="true" title="
    <div class='text-start'>
      {% if 'gnomad_mt_homoplasmic_frequency' in variant %}
        <div>
          <strong>gnomAD(MT) hom</strong>: {{ variant.gnomad_mt_homoplasmic_frequency|human_decimal if variant.gnomad_mt_homoplasmic_frequency }}
        </div>
      {% endif %}
      {% if 'gnomad_mt_heteroplasmic_frequency' in variant %}
        <div>
          <strong>gnomAD(MT) het</strong>: {{ variant.gnomad_mt_heteroplasmic_frequency|human_decimal if variant.gnomad_mt_heteroplasmic_frequency }}
        </div>
      {% endif %}

      {% if 'swegen_alu' in variant %}
        <div>
          <strong>SweGen ALU</strong>: {{ variant.swegen_alu|human_decimal if variant.swegen_alu }}
        </div>
      {% endif %}
      {% if 'swegen_herv' in variant %}
        <div>
          <strong>SweGen HERV</strong>: {{ variant.swegen_herv|human_decimal if variant.swegen_herv }}
        </div>
      {% endif %}
      {% if 'swegen_l1' in variant %}
        <div>
          <strong>SweGen L1</strong>: {{ variant.swegen_l1|human_decimal if variant.swegen_l1 }}
        </div>
      {% endif %}
      {% if 'swegen_sva' in variant %}
        <div>
          <strong>SweGen SVA</strong>: {{ variant.swegen_sva|human_decimal if variant.swegen_sva }}
        </div>
      {% endif %}

      {% if 'swegen' in variant %}
        <div>
          <strong>SweGen AF</strong>: {{ variant.swegen|human_decimal if variant.swegen }}
        </div>
      {% endif %}

      {% if 'gnomad_frequency' in variant %}
        <div>
          <strong>gnomAD</strong>: {{ variant.gnomad_frequency|human_decimal if variant.gnomad_frequency }}
        </div>
        <div>
          <strong>gnomAD (max)</strong>: {{ variant.max_gnomad_frequency|human_decimal if variant.max_gnomad_frequency }}
        </div>
      {% else %}
        <div>
          <strong>gnomAD</strong>: -
        </div>
      {% endif %}
      {% if 'thousand_genomes_frequency' in variant %}
      <div>
        <strong>1000G</strong>: {{ variant.thousand_genomes_frequency|human_decimal if variant.thousand_genomes_frequency }}
      </div>
      {% endif %}
      {% if 'max_thousand_genomes_frequency' in variant %}
      <div>
        <strong>1000G (max)</strong>: {{ variant.max_thousand_genomes_frequency|human_decimal if variant.max_thousand_genomes_frequency }}
      </div>
      {% endif %}
      {% if 'exac_frequency' in variant %}
      <div>
        <strong>ExAC</strong>:
        {{ variant.exac_frequency|human_decimal if variant.exac_frequency }}
      </div>
      {% endif %}
      {% if 'max_exac_frequency' in variant %}
      <div>
        <strong>ExAC (max)</strong>:
        {{ variant.max_exac_frequency|human_decimal if variant.max_exac_frequency }}
      </div>
      {% endif %}
    </div>
  ">
  <div>
  {% if variant.category == "mei" %}
    {{ variant.swegen_mei_max|human_decimal if variant.swegen_mei_max else '~' }}
  {% elif variant.chromosome != "MT" %}
    {{ variant.gnomad_frequency|human_decimal if variant.gnomad_frequency else '~' }}
  {% else %}
    {{ variant.gnomad_mt_homoplasmic_frequency|human_decimal if variant.gnomad_mt_homoplasmic_frequency else '~' }}
  {% endif %}
  </div>
  </div>
{% endmacro %}

{% macro external_stylesheets() %}
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.14.0-beta3/css/bootstrap-select.min.css" integrity="sha512-g2SduJKxa4Lbn3GW+Q7rNz+pKP9AWMR++Ta8fgwsZRCUsawjPvF/BxSMkGS61VsR9yinGoEgrHPGPn2mrj8+4w==" crossorigin="anonymous" referrerpolicy="no-referrer" />
{% endmacro %}

{% macro external_scripts() %}
  <script src="https://cdnjs.cloudflare.com/ajax/libs/dompurify/2.3.8/purify.min.js" integrity="sha512-M72KfQy4kPuLYC6CeTrN0eA17U1lXEMrr5qEJC/40CLdZGC3HpwPS0esQLqBHnxty2FIcuNdP9EqwSOCLEVJXQ==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.14.0-beta3/js/bootstrap-select.min.js" integrity="sha512-yrOmjPdp8qH8hgLfWpSFhC/+R9Cj9USL8uJxYIveJZGAiedxyIxwNw4RsLDlcjNlIRR4kkHaDHSmNHAkxFTmgg==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>

  <script src="https://cdn.datatables.net/1.12.0/js/jquery.dataTables.min.js" integrity="sha512-fu0WiDG5xqtX2iWk7cp17Q9so54SC+5lk/z/glzwlKFdEOwGG6piUseP2Sik9hlvlmyOJ0lKXRSuv1ltdVk9Jg==" referrerpolicy="no-referrer" crossorigin="anonymous"></script>
  <script src="https://cdn.datatables.net/1.12.0/js/dataTables.bootstrap5.min.js" integrity="sha512-nfoMMJ2SPcUdaoGdaRVA1XZpBVyDGhKQ/DCedW2k93MTRphPVXgaDoYV1M/AJQLCiw/cl2Nbf9pbISGqIEQRmQ==" referrerpolicy="no-referrer" crossorigin="anonymous"></script>
{% endmacro %}
